# Indice

- ⚡ [0.- Introducción](#0--introducción)
- 📌 [1.- Información general del dataset](#1--información-general-del-dataset)
- 📦 [2.- Instalar librerías necesarias](#2--instalar-librerías-necesarias)
- 📥 [3.- Descripción y Carga de dataset](#3--descripción-y-carga-de-dataset)
  - [3.1.- Carga del dataset](#31--carga-del-dataset)
  - [3.2.- Descripción del dataset](#32--descripción-del-dataset)
- 🗒️ [4.- Primer vistazo de datos](#4--primer-vistazo-de-datos)
  - [Tipos de cada atributo](#tipos-de-cada-atributo)
  - [Data numérica](#data-numérica)
  - [Porcentaje de valores NaN en las columnas](#porcentaje-de-valores-nan-en-las-columnas)
- 🔍 [5.- EDA (Exploratory Data Analysis) Análisis Univariante](#5--eda-exploratory-data-analysis-análisis-univariante)
  - [5.1.- Loan amount (Monto del préstamo)](#51--loan-amount-monto-del-préstamo)
  - [5.2.- Annual Income (Ingresos Anuales)](#52--annual-income-ingresos-anuales)
  - [5.3.- Interest Rate (Tipo de Interés)](#53--interest-rate-tipo-de-interés)
  - [5.4.- FICO scores (Puntuación de crédito FICO)](#54--fico-scores-puntuación-de-crédito-fico)
  - [5.5.- Purpose y Term (Objeto y Duración)](#55--purpose-y-term-objeto-y-duración)
  - [5.6.- Grade y Subgrades (Grados y Subgrados)](#56--grade-y-subgrades-grados-y-subgrados)
  - [5.7.- emp_length, home ownership, verification status y not verified status](#57--emp_length-home-ownership-verification-status-y-not-verified-status)
  - [5.8.- initial_list_status, loan_status y application_type](#58--initial_list_status-loan_status-y-application_type)
  - [5.9.- addr_state y emp_title](#59--addr_state-y-emp_title)
  - [5.10.- Hardship y related](#510--hardship-y-related)
- 🔍 [6.- EDA (Exploratory Data Analysis) Análisis Bivariante](#6--eda-exploratory-data-analysis-análisis-bivariante)
  - [6.1.- Loan status y relacionados](#61--loan-status-y-relacionados)
    - [6.1.1.- loan_status y term](#611--loan_status-y-term)
    - [6.1.2.- loan_status y grade](#612--loan_status-y-grade)
    - [6.1.3.- loan_status y home_ownership](#613--loan_status-y-home_ownership)
    - [6.1.4.- loan_status y verification_status](#614--loan_status-y-verification_status)
    - [6.1.5.- loan_status con annual_inc_range, emp_length, y purpose](#615--loan_status-con-annual_inc_range-emp_length-y-purpose)
    - [6.1.6.- loan_status y open_acc/ total_acc](#616--loan_status-y-open_acc-total_acc)
    - [6.1.7.- loan_status y pub_rec/ pub_rec_bankruptcies](#617--loan_status-y-pub_rec-pub_rec_bankruptcies)
    - [6.1.8.- Loan status y interest rate / loan amount](#618--loan-status-y-interest-rate-loan-amount)
    - [6.1.9.- Loan status per state](#619--loan-status-per-state)
  - [6.2.- Loan amount](#62--loan-amount)
    - [6.2.1.- Loan amount vs year experience](#621--loan-amount-vs-year-experience)
    - [6.2.2.- Loan amount y term](#622--loan-amount-y-term)
    - [6.2.3.- Loan amount y purpose](#623--loan-amount-y-purpose)
  - [6.3.- Purpose](#63--purpose)
    - [6.3.1.- Purpose y term](#631--purpose-y-term)
    - [6.3.2.- Purpose y grade](#632--purpose-y-grade)
  - [6.4.- Interest Rate](#64--interest-rate)
    - [6.4.1.- Interest rate y annual income](#641--interest-rate-y-annual-income)
    - [6.4.2.- Interest rate y term](#642--interest-rate-y-term)
    - [6.4.3.- Interest vs sub-grade](#643--interest-vs-sub-grade)

# ⚡ 0.- Introduccion

En este documento, se realiza un EDA univariante y bivariante del siguiente dataset con los datos de entre 2007 y 2020:

Loan_status_2007-2020Q3.csv

El presente documento ha sido elaborado con el fin de:

- Poder realizar un análisis extenso del dataset de Lending Club

- Sacar conclusiones que han ayudado a extraer los datos que creo que han sido más relevantes para el documento principal del repositorio:

1_Prediccion_Riesgo_Crediticio_Lending_Club_2019_7_modelos.ipynb

# 📌 1.- Informacion general del dataset

Data is collected originally from LendingClub data website. The LendingClub Notes platform retired on December 31, 2020 (https://www.lendingclub.com/investing/peer-to-peer), so the accuracy of this data sample solely depends on the collection of individuals on Kaggle.

Content: The personal information of customers is collected during the membership, application process and from credit bureaus such as date of birth, income, account information, credit history and credit score, credit card and other debt information.

Overall information of the dataset:

Source: https://www.kaggle.com/ethon0426/lending-club-20072020q1?select=Loan_status_2007-2020Q3.gzip, the zipped file contains a *.gzip file of data and an excel file of feature description.
Duration: 2007-2020
Context: data is collected from customers in US (https://help.lendingclub.com/hc/en-us/articles/213706208-Qualifying-for-a-personal-loan) ?
Currency: USD
Types of features:
Loan Characteristics such as loan amount, term and purpose
Demographic data such as employment status, employment length and house ownership
Behavioral data related to historical payments
Because the problem to solve here is prediction for loan outcomes before it is granted, so the features provided before a loan is granted will be focus on, and the rest will be ignored.

# 📦 2.- Instalar librerias necesarias

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
import seaborn as sns

import os

# 📥 3.- Descripcion y Carga de dataset

En este documento, se utiliza los siguientes documentos:

1) Loan_status_2007-2020Q3.gzip        (posteriormente .csv)

El dataset en formato gzip de 1,73 GB aproximadamente. Este doc hay que pasar a csv

2) LCDataDictionary.xlsx

Un doc excel en el que se proporciona información acerca del dataset.


Ambos archivos se han guardado en la carpeta Data de este repositorio.

## 3.1.- Carga del dataset

En esta sección, se va a proceder a extraer la información del siguiente doc:

Loan_status_2007-2020Q3.gzip

Con el fin de poder leer el archivo, tras realizar un par de pruebas, basta con cambiar la extensión de .gzip a .csv

Por ahora, vamos a convertir el doc gzip en csv.

Loan_status_2007-2020Q3.csv

In [2]:
data = pd.read_csv("./Data/Loan_status_2007-2020Q3.csv")

C:\Users\Usuario\AppData\Local\Temp\ipykernel_17264\2297645106.py:1: DtypeWarning: Columns (1,48,58,117,127,128,129,132,133,134,137) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("./Data/Loan_status_2007-2020Q3.csv")


In [3]:
data.shape

(2925493, 142)

El dataset inicial está compuesto por:
- 2925493 rows
- 142 columns

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925493 entries, 0 to 2925492
Columns: 142 entries, Unnamed: 0 to debt_settlement_flag
dtypes: float64(106), int64(1), object(35)
memory usage: 3.1+ GB


Leemos el dataset

Aquí es interesante aplicar la extensión Data Wrangler en el df para ver sus propiedades

In [5]:
data

Unnamed: 0         id  loan_amnt  funded_amnt  funded_amnt_inv  \
0                 0    1077501     5000.0       5000.0           4975.0   
1                 1    1077430     2500.0       2500.0           2500.0   
2                 2    1077175     2400.0       2400.0           2400.0   
3                 3    1076863    10000.0      10000.0          10000.0   
4                 4    1075358     3000.0       3000.0           3000.0   
...             ...        ...        ...          ...              ...   
2925488      105446  102556443    24000.0      24000.0          24000.0   
2925489      105447  102653304    10000.0      10000.0          10000.0   
2925490      105448  102628603    10050.0      10050.0          10050.0   
2925491      105449  102196576     6000.0       6000.0           6000.0   
2925492      105450   99799684    30000.0      30000.0          30000.0   

               term int_rate  installment grade sub_grade  ...  \
0         36 months   10.65%       162.87     B        B2  ...   
1         60 months   15.27%        59.83     C        C4  ...   
2         36 months   15.96%        84.33     C        C5  ...   
3         36 months   13.49%       339.31     C        C1  ...   
4         60 months   12.69%        67.79     B        B5  ...   
...             ...      ...          ...   ...       ...  ...   
2925488   60 months   23.99%       690.30     E        E2  ...   
2925489   36 months    7.99%       313.32     A        A5  ...   
2925490   36 months   16.99%       358.26     D        D1  ...   
2925491   36 months   11.44%       197.69     B        B4  ...   
2925492   60 months   25.49%       889.18     E        E4  ...   

        hardship_start_date hardship_end_date payment_plan_start_date  \
0                       NaN               NaN                     NaN   
1                       NaN               NaN                     NaN   
2                       NaN               NaN                     NaN   
3                       NaN               NaN                     NaN   
4                       NaN               NaN                     NaN   
...                     ...               ...                     ...   
2925488                 NaN               NaN                     NaN   
2925489                 NaN               NaN                     NaN   
2925490                 NaN               NaN                     NaN   
2925491                 NaN               NaN                     NaN   
2925492                 NaN               NaN                     NaN   

         hardship_length hardship_dpd hardship_loan_status  \
0                    NaN          NaN                  NaN   
1                    NaN          NaN                  NaN   
2                    NaN          NaN                  NaN   
3                    NaN          NaN                  NaN   
4                    NaN          NaN                  NaN   
...                  ...          ...                  ...   
2925488              NaN          NaN                  NaN   
2925489              NaN          NaN                  NaN   
2925490              NaN          NaN                  NaN   
2925491              NaN          NaN                  NaN   
2925492              NaN          NaN                  NaN   

        orig_projected_additional_accrued_interest  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   
...                                            ...   
2925488                                        NaN   
2925489                                        NaN   
2925490                                        NaN   
2925491                                        NaN   
2925492                                        NaN   

        hardship_payoff_balance_amount hards

## 3.2.- Descripcion del dataset

En esta sección, se va a proceder a extraer la información del siguiente doc:

LCDataDictionary.xlsx

Este archivo en formato excel, contiene la descripción de los índices de cada una de las 142 columnas del archivo.

El archivo excel, traspone las 142 columnas a 142 filas, y muestra una breve descripción de cada uno de los índices.

In [6]:
description = pd.read_excel("./Data/LCDataDictionary.xlsx").dropna(axis=1, how='all', inplace=False)

In [7]:
description

LoanStatNew                                        Description
0           acc_now_delinq  The number of accounts on which the borrower i...
1     acc_open_past_24mths         Number of trades opened in past 24 months.
2               addr_state  The state provided by the borrower in the loan...
3                 all_util              Balance to credit limit on all trades
4               annual_inc  The self-reported annual income provided by th...
..                     ...                                                ...
138    verification_status  Indicates if income was verified by LC, not ve...
139  verified_status_joint  Indicates if the co-borrowers' joint income wa...
140               zip_code  The first 3 numbers of the zip code provided b...
141                    NaN                                                NaN
142                    NaN  * Employer Title replaces Employer Name for al...

[143 rows x 2 columns]

# 🗒️ 4.- Primer vistazo de datos

## Tipos de cada atributo

En primer lugar, vamos a ver los tipos de cada atributo:

check the types of each attribute

In [8]:
data.dtypes.sort_values()

Unnamed: 0                    int64
num_bc_tl                   float64
num_bc_sats                 float64
num_actv_rev_tl             float64
num_actv_bc_tl              float64
                             ...   
issue_d                      object
sec_app_earliest_cr_line     object
verification_status          object
emp_length                   object
debt_settlement_flag         object
Length: 142, dtype: object

While most the features are numerical, some has object type which should be examined further in the next part. Due to the large number of variables, we analysis some features which is related to the problem stated.

## Data numerica

Análisis descriptivo

In [9]:
#data.describe().transpose()

Vamos a ver de forma completo sólo para este caso:

In [10]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(data.describe().transpose())

count           mean  \
Unnamed: 0                                  2925493.0   86974.194940   
loan_amnt                                   2925492.0   15358.775105   
funded_amnt                                 2925492.0   15354.704909   
funded_amnt_inv                             2925492.0   15340.046287   
installment                                 2925492.0     452.391358   
annual_inc                                  2925488.0   79937.274015   
dti                                         2922384.0      19.296480   
delinq_2yrs                                 2925463.0       0.289701   
fico_range_low                              2925492.0     700.511030   
fico_range_high                             2925492.0     704.511235   
inq_last_6mths                              2925462.0       0.554213   
mths_since_last_delinq                      1388990.0      34.949602   
mths_since_last_record                       427448.0      74.706018   
open_acc                                    2925463.0      11.676163   
pub_rec                                     2925463.0       0.176285   
revol_bal                                   2925492.0   16964.840934   
total_acc                                   2925463.0      24.018350   
out_prncp                                   2925492.0    4038.696551   
out_prncp_inv                               2925492.0    4038.123945   
total_pymnt                                 2925492.0   12574.550469   
total_pymnt_inv                             2925492.0   12559.755026   
total_rec_prncp                             2925492.0    9917.802237   
total_rec_int                               2925492.0    2484.796896   
total_rec_late_fee                          2925492.0       1.764549   
recoveries                                  2925492.0     170.186755   
collection_recovery_fee                     2925492.0      29.099492   
last_pymnt_amnt                             2925492.0    3479.507389   
last_fico_range_high                        2925492.0     690.373418   
last_fico_range_low                         2925492.0     677.845552   
collections_12_mths_ex_med                  2925347.0       0.017682   
mths_since_last_major_derog                  722917.0      44.601649   
policy_code                                 2925492.0       1.000000   
annual_inc_joint                             210513.0  130581.834573   
dti_joint                                    210509.0      19.421121   
acc_now_delinq                              2925463.0       0.003230   
tot_coll_amt                                2855216.0     216.882722   
tot_cur_bal                                 2855216.0  145377.339755   
open_acc_6m                                 2059362.0       0.931240   
open_act_il                                 2059363.0       2.826787   
open_il_12m                                 2059363.0       0.698881   
open_il_24m                                 2059363.0       1.603937   
mths_since_rcnt_il                          1998963.0      20.403080   
total_bal_il                                2059363.0   36835.366107   
il_util                                     1768344.0      68.963711   
open_rv_12m                                 2059363.0       1.253812   
open_rv_24m                                 2059363.0       2.645740   
max_bal_bc                                  2059363.0    6018.203809   
all_util                                    2059007.0      56.278664   
total_rev_hi_lim                            2855216.0   36338.643163   
inq_fi                                      2059363.0       1.071141   
total_cu_tl                                 2059362.0       1.517789   
inq_last_12m                                2059362.0       2.001608   
acc_open_past_24mths                        2875462.0       4.490981   
avg_cur_bal                                 2855104.0   13712.209752   
bc_open_to_buy                              2842824.0   12682.044322   
bc_util         

Tras analizar la tabla previa, se va a mostrar algunos puntos clave basados en las columnas y sus estadísticas descriptivas.

Estos puntos proporcionan una comprensión más profunda y detallada de las estadísticas descriptivas y las características de los datos en las columnas:

1) Count (Recuento)

- Consistencia de Datos: La mayoría de las columnas tienen un recuento cercano a 2.925.492. lo que sugiere que los datos están bastante completos. Sin embargo, algunas columnas tienen recuentos significativamente menores.

- Datos Faltantes: La diferencia en el recuento de las columnas puede indicar la presencia de datos faltantes o no aplicables.

- Variables Específicas: Algunas columnas tienen recuentos muy bajos debido a su naturaleza específica. Por ejemplo, hardship_length tiene un recuento de 143,638, lo que sugiere que esta información solo se aplica a una pequeña fracción de los registros.

2) Mean (Media)

- Montos de Préstamo: La media del loan_amnt es $15,358.78, lo que representa el monto promedio de los préstamos otorgados. Este valor es útil para entender el tamaño típico de los préstamos.

- Ingresos Anuales: La media del annual_inc es $79,937.27, proporcionando una indicación del ingreso anual promedio de los prestatarios. Este valor puede ser influenciado por algunos valores atípicos muy altos.

- Relación Deuda-Ingreso: La media de dti es 19.30, lo que ayuda a evaluar la carga de deuda promedio en relación con los ingresos de los prestatarios.

3) Standard Deviation (Desviación Estándar)

- Variabilidad en Ingresos: La desviación estándar de annual_inc es $111,747.61, indicando una amplia variabilidad en los ingresos anuales de los prestatarios. Esto sugiere una población prestataria con ingresos muy diversos.

- Consistencia en Puntajes FICO: Las desviaciones estándar de fico_range_low (33.77) y fico_range_high (33.77) son relativamente bajas, sugiriendo que los puntajes FICO son bastante consistentes entre los prestatarios.

- Montos de Préstamo: La desviación estándar de loan_amnt es $9,478.38, indicando una considerable variabilidad en los montos de los préstamos, lo que refleja diferentes necesidades de financiamiento entre los prestatarios.

4) Min (Valor Mínimo)

- Valores Extremadamente Bajos: Varias columnas tienen un valor mínimo de 0, como funded_amnt_inv y annual_inc, lo que puede representar prestatarios sin ingresos reportados o valores faltantes.

- Posibles Errores en Datos: La columna dti tiene un valor mínimo de -1, lo que es inusual y puede indicar un error en los datos o un valor codificado para datos faltantes.

- Impacto en el Análisis: Los valores mínimos extremadamente bajos, como en installment ($4.93), pueden afectar el análisis de datos y las visualizaciones, especialmente si representan casos anómalos o errores en los datos.

5) 25% (Primer Cuartil)

- Montos de Préstamo Moderados: El 25% de los préstamos (loan_amnt) son de $8,000 o menos. Esto indica que una cuarta parte de los prestatarios están solicitando préstamos relativamente pequeños.

- Ingresos Anuales: El primer cuartil de los ingresos anuales (annual_inc) es de $47,000. Esto sugiere que el 25% de los prestatarios tienen ingresos anuales de $47,000 o menos, indicando un segmento de prestatarios con ingresos bajos a moderados.

- Relación Deuda-Ingreso: Para dti, el 25% de los prestatarios tienen una relación deuda-ingreso de 12.08 o menos, lo que sugiere una carga de deuda manejable para este grupo.

6) 50% (Mediana)

- Mediana del Monto del Préstamo: La mediana del loan_amnt es $13,000. Esto significa que la mitad de los préstamos son menores o iguales a $13,000, reflejando el monto de préstamo más común entre los prestatarios.

- Mediana de Ingresos Anuales: La mediana del ingreso anual (annual_inc) es $66,000, indicando que la mitad de los prestatarios tienen ingresos anuales menores o iguales a $66,000.

- Mediana de la Relación Deuda-Ingreso: La mediana del dti es 18.10, mostrando que el prestatario promedio tiene una relación deuda-ingreso de 18.10, lo cual es una métrica clave para evaluar la capacidad de pago de los prestatarios.

7) 75% (Tercer Cuartil)

- Montos de Préstamo Altos: El 75% de los préstamos (loan_amnt) son de $20,000 o menos. Esto sugiere que la mayoría de los préstamos no superan los $20,000, aunque hay un 25% que son superiores a este monto.

- Ingresos Anuales Altos: El tercer cuartil de los ingresos anuales (annual_inc) es $95,000, indicando que el 25% de los prestatarios tienen ingresos anuales superiores a este valor, lo que refleja un segmento de prestatarios con ingresos altos.

- Relación Deuda-Ingreso: Para dti, el 75% de los prestatarios tienen una relación deuda-ingreso de 24.88 o menos. Esto muestra que, aunque una mayoría tiene una carga de deuda manejable, hay un 25% con una relación deuda-ingreso superior a este valor, lo cual puede representar un mayor riesgo financiero.

8) Max (Valor Máximo)

- Valores Atípicos Significativos: Columnas como annual_inc tienen valores máximos extremadamente altos ($110,000,000), lo que sugiere la presencia de valores atípicos que pueden distorsionar el análisis si no se manejan adecuadamente.

- Rango de Montos: El loan_amnt tiene un valor máximo de $40,000, mostrando el límite superior de los préstamos otorgados, lo cual es útil para comprender el rango de financiación.

- Cargas Financieras: La columna revol_bal tiene un valor máximo de $2,904,836, indicando que algunos prestatarios tienen deudas revolving extremadamente altas, lo que podría representar un alto riesgo de crédito.

Los anteriores puntos, nos puede dar pistas a la hora de escoger datos que queramos analizar con más detalle en el EDA univariante y bivariante

## Porcentaje de valores NaN en las columnas

Comprobar y calcular el porcentaje de valores nulos sobre el total de valores contabilizados en cada columna

In [ ]:
null_values = data.isnull().sum()

len(df.index) sirve para contar el número total de valores de una columna que se supone que tiene

In [ ]:
percent_nullvalues = null_values/len(data.index)*100

Ordenar el porcentaje de valores perdidos de mayor a menor

In [ ]:
percent_nullvalues.sort_values(ascending=False, axis=0)

# 🔍 5.- EDA (Exploratory Data Analysis) Analisis Univariante

Los objetivos de este EDA univariante van a ser los siguientes:

1) Analizar el Loan Ammount

2) Analizar el Annual Income

3) Analizar el Interest Rate

4) Analizar los FICO scores

5) Analizar los Purpose y Terms

6) Analizar los Grades y los Subgrades

7) Analizar los emp_length , home ownership , verification status y not verified status

8) Analizar los initial_list_status, loan_status y application_type

9) Analizar los addr_state y emp_title

10) Analizar los Hardship y relacionados

Esto servirá para comprender qué variables son importantes, ver estadísticas resumidas y visualizar los datos.

## 5.1.- Loan ammount (Monto del prestamo)

Crear bins para loan_amnt_range

In [11]:
bins = list(range(0, 45000, 5000))
labels = [f'{i}-{i+5000}' for i in range(0, 40000, 5000)]

In [12]:
data['loan_amnt_range_5000'] = pd.cut(data['loan_amnt'], bins=bins, labels=labels, right=False)

In [13]:
loan_amnt_range_counts = data['loan_amnt_range_5000'].value_counts().sort_index()

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=loan_amnt_range_counts.index.astype(str),
    y=loan_amnt_range_counts.values,
    marker=dict(color=loan_amnt_range_counts.values, colorscale='Viridis'),
    text=loan_amnt_range_counts.values,
    textposition='auto'
))

fig.update_layout(
    title='Distribution of Loan amount (in $5000 ranges)',
    xaxis_title='Loan Amount Range (USD)',
    yaxis_title='Frequency',
    title_font_size=16,
    width=1200,
    height=600,
    bargap=0.2  
)

fig.show()

Gráfico de Pastel de los Rangos del Monto del Préstamo

In [ ]:
fig, ax = plt.subplots(figsize=(10, 6))
data['loan_amnt_range_5000'].value_counts().sort_index().plot.pie(ax=ax, autopct='%.2f%%')
ax.set_title('Loan amount range (USD in $5000 ranges)', fontsize=16)
ax.set_ylabel("")  
plt.show()

## 5.2.- Annual Income (Ingresos Anuales)

In [16]:
annual_income_bin = data['annual_inc'].value_counts(bins=30).sort_values()
annual_income_bin.tail(20)

(22000000.0, 25666666.667]              0
(62333333.333, 66000000.0]              0
(69666666.667, 73333333.333]            0
(73333333.333, 77000000.0]              0
(55000000.0, 58666666.667]              0
(80666666.667, 84333333.333]            0
(84333333.333, 88000000.0]              0
(88000000.0, 91666666.667]              0
(91666666.667, 95333333.333]            0
(95333333.333, 99000000.0]              0
(99000000.0, 102666666.667]             0
(102666666.667, 106333333.333]          0
(66000000.0, 69666666.667]              0
(18333333.333, 22000000.0]              0
(25666666.667, 29333333.333]            0
(106333333.333, 110000000.0]            1
(58666666.667, 62333333.333]            1
(7333333.333, 11000000.0]              62
(3666666.667, 7333333.333]            195
(-110000.001, 3666666.667]        2925229
Name: count, dtype: int64

In [17]:
bins_annual_income = [0, 0.1, 20000, 40000, 60000, 80000, 100000, 120000, 140000, 160000, 180000, 200000, float('inf')]
labels_annual_income = [
    '0', '0.1-20000', '20000-40000', '40000-60000', 
    '60000-80000', '80000-100000', '100000-120000', 
    '120000-140000', '140000-160000', '160000-180000', 
    '180000-200000', '200000+'
]

In [18]:
data['annual_inc_range'] = pd.cut(data['annual_inc'], bins=bins_annual_income, labels=labels_annual_income, right=False, include_lowest=True)

In [19]:
annual_inc_range_counts = data['annual_inc_range'].value_counts().sort_index()

In [20]:
total_count = annual_inc_range_counts.sum()
percentage_labels = (annual_inc_range_counts / total_count * 100).round(2).astype(str) + '%'

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=annual_inc_range_counts.index.astype(str),
    y=annual_inc_range_counts.values,
    marker=dict(color=annual_inc_range_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in annual_inc_range_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(annual_inc_range_counts.index.astype(str), annual_inc_range_counts.values, percentage_labels)):
    fig.add_annotation(
        x=x,
        y=y,
        text=percentage,
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Annual Income',
    xaxis_title='Annual Income Range (USD)',
    yaxis_title='Frequency',
    title_font_size=16,
    width=1200,
    height=600,
    bargap=0.2  
)

fig.show()

## 5.3.- Interest Rate (Tipo de Interes)

In [22]:
data['int_rate'] = pd.to_numeric(data['int_rate'].str.replace('%', ''))

In [23]:
bin_interest = data['int_rate'].value_counts(bins=20).sort_values(ascending=True)
bin_interest

(27.138, 28.422]                 5696
(29.706, 30.99]                  9667
(25.854, 27.138]                14076
(28.422, 29.706]                14804
(23.286, 24.57]                 28017
(24.57, 25.854]                 34915
(22.002, 23.286]                44599
(20.718, 22.002]                52044
(19.434, 20.718]                98152
(18.15, 19.434]                113839
(5.2829999999999995, 6.594]    167140
(15.582, 16.866]               178777
(16.866, 18.15]                200024
(6.594, 7.878]                 241097
(14.298, 15.582]               241359
(9.162, 10.446]                251173
(13.014, 14.298]               286514
(11.73, 13.014]                299071
(7.878, 9.162]                 301361
(10.446, 11.73]                343167
Name: count, dtype: int64

In [24]:
data['int_rate'].head()

0    10.65
1    15.27
2    15.96
3    13.49
4    12.69
Name: int_rate, dtype: float64

In [25]:
bins_interest = [0, 0.01, 3, 5, 8, 12, 16, 20, float('inf')]
labels_interest = [
    '0', '0.01-3', '3-5', '5-8', 
    '8-12', '12-16', '16-20', '20+'
]

In [26]:
data['interest_range'] = pd.cut(data['int_rate'], bins=bins_interest, labels=labels_interest, right=False, include_lowest=True)

In [27]:
interest_range_counts = data['interest_range'].value_counts().sort_index()

In [28]:
total_count = interest_range_counts.sum()
percentage_labels = (interest_range_counts / total_count * 100).round(2).astype(str) + '%'

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=interest_range_counts.index.astype(str),
    y=interest_range_counts.values,
    marker=dict(color=interest_range_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in interest_range_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(interest_range_counts.index.astype(str), interest_range_counts.values, percentage_labels)):
    fig.add_annotation(
        x=x,
        y=y,
        text=percentage,
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Interest Rates',
    xaxis_title='Interest Rate Range (%)',
    yaxis_title='Frequency',
    title_font_size=16,
    width=1200,
    height=500,
    bargap=0.2  
)

fig.show()

## 5.4.- FICO scores (Puntuacion de credito FICO)

Enlaces interesantes FICO:

https://us.as.com/actualidad/como-funciona-el-fico-score-cual-es-un-buen-puntaje-de-credito-n/

https://core.ac.uk/download/pdf/6538561.pdf

Hay 2 variables con el contenido similar es fico_range_low y fico_range_high, por lo que se necesita una transformación.

In [30]:
data['diff'] = data['fico_range_high'] - data['fico_range_low']
data['diff'].head()

0    4.0
1    4.0
2    4.0
3    4.0
4    4.0
Name: diff, dtype: float64

In [31]:
data['diff'].tail()

2925488    4.0
2925489    4.0
2925490    4.0
2925491    4.0
2925492    4.0
Name: diff, dtype: float64

In [32]:
average_diff = round(data['diff'].mean(), 2)
average_diff

4.0

In [33]:
data['diff'].value_counts()

diff
4.0    2924892
5.0        600
Name: count, dtype: int64

Efectivamente la media entre la variable FICO low y high en general es de 4. Por lo que tomaremos las cifras medias para el análisis posterior.

In [34]:
data['FICO_average'] = (data['fico_range_high'] + data['fico_range_low'])/2
data['FICO_average'].head()

0    737.0
1    742.0
2    737.0
3    692.0
4    697.0
Name: FICO_average, dtype: float64

In [35]:
bins_fico = list(range(500, 925, 25))
labels_fico = [f'{i}-{i+25}' for i in range(500, 900, 25)] + ['900+']

In [36]:
data['FICO_range'] = pd.cut(data['FICO_average'], bins=bins_fico + [float('inf')], labels=labels_fico, right=False, include_lowest=True)

In [37]:
fico_range_counts = data['FICO_range'].value_counts().sort_index()

In [38]:
total_count = fico_range_counts.sum()
percentage_labels = (fico_range_counts / total_count * 100).round(2).astype(str) + '%'

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=fico_range_counts.index.astype(str),
    y=fico_range_counts.values,
    marker=dict(color=fico_range_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in fico_range_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(fico_range_counts.index.astype(str), fico_range_counts.values, percentage_labels)):
    fig.add_annotation(
        x=x,
        y=y,
        text=percentage,
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of FICO Scores',
    xaxis_title='FICO Score Range',
    yaxis_title='Frequency',
    title_font_size=16,
    width=1200,
    height=500,
    bargap=0.2  
)

fig.show()

In [40]:
bins_fico_2 = [0, 580, 670, 740, 800, float('inf')]
labels_fico_2 = ['<580', '580-669', '670-739', '740-799', '800+']

In [41]:
data['FICO_range_2'] = pd.cut(data['FICO_average'], bins=bins_fico_2, labels=labels_fico_2, right=False, include_lowest=True)

fico_range_counts = data['FICO_range_2'].value_counts().sort_index()

total_count = fico_range_counts.sum()
percentage_labels = (fico_range_counts / total_count * 100).round(2).astype(str) + '%'

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=fico_range_counts.index.astype(str),
    y=fico_range_counts.values,
    marker=dict(color=fico_range_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in fico_range_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(fico_range_counts.index.astype(str), fico_range_counts.values, percentage_labels)):
    fig.add_annotation(
        x=x,
        y=y,
        text=percentage,
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of FICO scores with reference values',
    xaxis_title='FICO Score Range',
    yaxis_title='Frequency',
    title_font_size=16,
    width=900,
    height=600,
    bargap=0.2  
)

fig.show()

## 5.5.- Purpose y Term (Objeto y Duracion)

Las características se presentan lado a lado, pero es para una disposición óptima, no para comprobar su relación mutua.

Purpose

In [43]:
data['purpose'].head(30)

0            credit_card
1                    car
2         small_business
3                  other
4                  other
5                wedding
6     debt_consolidation
7                    car
8         small_business
9                  other
10    debt_consolidation
11    debt_consolidation
12    debt_consolidation
13           credit_card
14                 other
15    debt_consolidation
16      home_improvement
17        major_purchase
18               medical
19    debt_consolidation
20    debt_consolidation
21    debt_consolidation
22           credit_card
23    debt_consolidation
24        major_purchase
25           credit_card
26    debt_consolidation
27                 other
28    debt_consolidation
29           credit_card
Name: purpose, dtype: object

In [44]:
purpose_counts = data['purpose'].value_counts(normalize=True).mul(100)
purpose_values = data['purpose'].value_counts()

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=purpose_counts.index,
    y=purpose_counts.values,
    marker=dict(color=purpose_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in purpose_values.values],  
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(purpose_counts.index, purpose_counts.values, purpose_counts)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Loan purposes',
    xaxis_title='Purpose',
    yaxis_title='Percentage (%)',
    title_font_size=16,
    width=900,
    height=600,
    bargap=0.2  
)

fig.show()

Term

In [ ]:
term_counts = data['term'].value_counts()

fig_pie = px.pie(
    data_frame=data,
    names='term',
    title='Loan term',
    color_discrete_sequence=px.colors.sequential.Plasma,  
    hole=0.3  
)

fig_pie.update_layout(
    width=700,
    height=500
)

fig_pie.show()

## 5.6.- Grade y Subgrades (Grados y Subgrados)

In [47]:
unique_grades = data['grade'].unique()
unique_grades

array(['B', 'C', 'A', 'E', 'F', 'D', 'G', nan], dtype=object)

In [48]:
unique_subgrades = data['sub_grade'].unique()
unique_subgrades

array(['B2', 'C4', 'C5', 'C1', 'B5', 'A4', 'E1', 'F2', 'C3', 'B1', 'D1',
       'A1', 'B3', 'B4', 'C2', 'D2', 'A3', 'A5', 'D5', 'A2', 'E4', 'D3',
       'D4', 'F3', 'E3', 'F4', 'F1', 'E5', 'G4', 'E2', 'G3', 'G2', 'G1',
       'F5', 'G5', nan], dtype=object)

Grade

In [49]:
grade_counts = data['grade'].value_counts().sort_index()
grade_counts

grade
A    656302
B    857573
C    802280
D    416280
E    139036
F     41836
G     12185
Name: count, dtype: int64

In [50]:
total_count_grades = grade_counts.sum()
percentage_labels_grades = (grade_counts / total_count_grades * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=grade_counts.index,
    y=grade_counts.values,
    marker=dict(color=grade_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in grade_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(grade_counts.index, grade_counts.values, percentage_labels_grades)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Grades',
    xaxis_title='Grade',
    yaxis_title='Frequency',
    title_font_size=16,
    width=1200,
    height=500,
    bargap=0.2  
)

fig.show()

Subgrade

In [52]:
sub_grade_counts = data['sub_grade'].value_counts().sort_index()

In [53]:
total_count_subgrades = sub_grade_counts.sum()
percentage_labels_subgrades = (sub_grade_counts / total_count_subgrades * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=sub_grade_counts.index,
    y=sub_grade_counts.values,
    marker=dict(color=sub_grade_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in sub_grade_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(sub_grade_counts.index, sub_grade_counts.values, percentage_labels_subgrades)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Sub-Grades',
    xaxis_title='Sub-Grade',
    yaxis_title='Frequency',
    title_font_size=16,
    width=1500,
    height=600,
    bargap=0.2  
)

fig.show()

## 5.7.- emp_length , home ownership , verification status y not verified status

emp_length

In [55]:
unique_emp_lengths = data['emp_length'].unique()
unique_emp_lengths

array(['10+ years', '< 1 year', '1 year', '3 years', '8 years', '9 years',
       '4 years', '5 years', '6 years', '2 years', '7 years', nan],
      dtype=object)

In [56]:
emp_length_counts = data['emp_length'].value_counts().sort_index()

total_count_emp_length = emp_length_counts.sum()
percentage_labels_emp_length = (emp_length_counts / total_count_emp_length * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=emp_length_counts.index,
    y=emp_length_counts.values,
    marker=dict(color=emp_length_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in emp_length_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(emp_length_counts.index, emp_length_counts.values, percentage_labels_emp_length)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Employment Length',
    xaxis_title='Employment Length',
    yaxis_title='Frequency',
    title_font_size=16,
    width=1200,
    height=500,
    bargap=0.2  
)

fig.show()

home ownership

In [58]:
home_ownership_counts = data['home_ownership'].value_counts().sort_index()

total_count_home_ownership = home_ownership_counts.sum()
percentage_labels_home_ownership = (home_ownership_counts / total_count_home_ownership * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=home_ownership_counts.index,
    y=home_ownership_counts.values,
    marker=dict(color=home_ownership_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in home_ownership_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(home_ownership_counts.index, home_ownership_counts.values, percentage_labels_home_ownership)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Home Ownership',
    xaxis_title='Home Ownership',
    yaxis_title='Frequency',
    title_font_size=16,
    width=900,
    height=500,
    bargap=0.2  
)

fig.show()

verification status

In [60]:
verification_status_counts = data['verification_status'].value_counts().sort_index()

total_count_verification_status = verification_status_counts.sum()
percentage_labels_verification_status = (verification_status_counts / total_count_verification_status * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=verification_status_counts.index,
    y=verification_status_counts.values,
    marker=dict(color=verification_status_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in verification_status_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(verification_status_counts.index, verification_status_counts.values, percentage_labels_verification_status)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Verification Status',
    xaxis_title='Verification Status',
    yaxis_title='Frequency',
    title_font_size=16,
    width=900,
    height=400,
    bargap=0.2  
)

fig.show()

## 5.8.- initial_list_status, loan_status y application_type

initial_list_status

In [62]:
initial_list_status_counts = data['initial_list_status'].value_counts().sort_index()

total_count_initial_list_status = initial_list_status_counts.sum()
percentage_labels_initial_list_status = (initial_list_status_counts / total_count_initial_list_status * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=initial_list_status_counts.index,
    y=initial_list_status_counts.values,
    marker=dict(color=initial_list_status_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in initial_list_status_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(initial_list_status_counts.index, initial_list_status_counts.values, percentage_labels_initial_list_status)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Initial List Status',
    xaxis_title='Initial List Status',
    yaxis_title='Frequency',
    title_font_size=16,
    width=900,
    height=400,
    bargap=0.2  
)

fig.show()

loan_status

In [64]:
loan_status_counts = data['loan_status'].value_counts().sort_index()

total_count_loan_status = loan_status_counts.sum()
percentage_labels_loan_status = (loan_status_counts / total_count_loan_status * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=loan_status_counts.index,
    y=loan_status_counts.values,
    marker=dict(color=loan_status_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in loan_status_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(loan_status_counts.index, loan_status_counts.values, percentage_labels_loan_status)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Loan Status',
    xaxis_title='Loan Status',
    yaxis_title='Frequency',
    title_font_size=16,
    width=1200,
    height=600,
    bargap=0.2  
)

fig.show()

application_type

In [66]:
application_type_counts = data['application_type'].value_counts().sort_index()

total_count_application_type = application_type_counts.sum()
percentage_labels_application_type = (application_type_counts / total_count_application_type * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=application_type_counts.index,
    y=application_type_counts.values,
    marker=dict(color=application_type_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in application_type_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(application_type_counts.index, application_type_counts.values, percentage_labels_application_type)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Application Type',
    xaxis_title='Application Type',
    yaxis_title='Frequency',
    title_font_size=16,
    width=900,
    height=400,
    bargap=0.2 
)

fig.show()

## 5.9.- addr_state y emp_title

addr_state

In [68]:
addr_state_counts = data['addr_state'].value_counts().sort_values(ascending=False)

total_count_addr_state = addr_state_counts.sum()
percentage_labels_addr_state = (addr_state_counts / total_count_addr_state * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=addr_state_counts.index,
    y=addr_state_counts.values,
    marker=dict(color=addr_state_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in addr_state_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(addr_state_counts.index, addr_state_counts.values, percentage_labels_addr_state)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Address State',
    xaxis_title='Address State',
    yaxis_title='Frequency',
    title_font_size=16,
    width=1200,
    height=800,
    bargap=0.2  
)

fig.show()

emp_title

In [70]:
emp_title_counts = data['emp_title'].value_counts().sort_values(ascending=False).head(20)  

total_count_emp_title = emp_title_counts.sum()
percentage_labels_emp_title = (emp_title_counts / total_count_emp_title * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=emp_title_counts.index,
    y=emp_title_counts.values,
    marker=dict(color=emp_title_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in emp_title_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(emp_title_counts.index, emp_title_counts.values, percentage_labels_emp_title)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Top 20 Employment Titles',
    xaxis_title='Employment Title',
    yaxis_title='Frequency',
    title_font_size=16,
    width=1200,
    height=800,
    bargap=0.2  
)

fig.show()

## 5.10.- Hardship y related

hardship_type

In [72]:
hardship_type_counts = data['hardship_type'].value_counts().sort_values(ascending=False)

total_count_hardship_type = hardship_type_counts.sum()
percentage_labels_hardship_type = (hardship_type_counts / total_count_hardship_type * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=hardship_type_counts.index,
    y=hardship_type_counts.values,
    marker=dict(color=hardship_type_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in hardship_type_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(hardship_type_counts.index, hardship_type_counts.values, percentage_labels_hardship_type)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Hardship Types',
    xaxis_title='Hardship Type',
    yaxis_title='Frequency',
    title_font_size=16,
    width=900,
    height=600,
    bargap=0.2  
)

fig.show()

hardship_reason

In [74]:
hardship_reason_counts = data['hardship_reason'].value_counts().sort_values(ascending=False)

total_count_hardship_reason = hardship_reason_counts.sum()
percentage_labels_hardship_reason = (hardship_reason_counts / total_count_hardship_reason * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=hardship_reason_counts.index,
    y=hardship_reason_counts.values,
    marker=dict(color=hardship_reason_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in hardship_reason_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(hardship_reason_counts.index, hardship_reason_counts.values, percentage_labels_hardship_reason)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Hardship Reasons',
    xaxis_title='Hardship Reason',
    yaxis_title='Frequency',
    title_font_size=16,
    width=900,
    height=600,
    bargap=0.2  
)

fig.show()

hardship_flag

In [76]:
hardship_flag_counts = data['hardship_flag'].value_counts().sort_values(ascending=False)

total_count_hardship_flag = hardship_flag_counts.sum()
percentage_labels_hardship_flag = (hardship_flag_counts / total_count_hardship_flag * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=hardship_flag_counts.index,
    y=hardship_flag_counts.values,
    marker=dict(color=hardship_flag_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in hardship_flag_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(hardship_flag_counts.index, hardship_flag_counts.values, percentage_labels_hardship_flag)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Hardship Flag',
    xaxis_title='Hardship Flag',
    yaxis_title='Frequency',
    title_font_size=16,
    width=900,
    height=400,
    bargap=0.2 
)

fig.show()

hardship_status

In [78]:
hardship_status_counts = data['hardship_status'].value_counts().sort_values(ascending=False)

total_count_hardship_status = hardship_status_counts.sum()
percentage_labels_hardship_status = (hardship_status_counts / total_count_hardship_status * 100).round(2)

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=hardship_status_counts.index,
    y=hardship_status_counts.values,
    marker=dict(color=hardship_status_counts.values, colorscale='Viridis'),
    text=[f'{count}' for count in hardship_status_counts.values],
    textposition='inside',
    insidetextanchor='middle'
))

for i, (x, y, percentage) in enumerate(zip(hardship_status_counts.index, hardship_status_counts.values, percentage_labels_hardship_status)):
    fig.add_annotation(
        x=x,
        y=y,
        text=f'{percentage:.2f}%',
        showarrow=False,
        yshift=10,
        font=dict(size=12)
    )

fig.update_layout(
    title='Distribution of Hardship Status',
    xaxis_title='Hardship Status',
    yaxis_title='Frequency',
    title_font_size=16,
    width=900,
    height=400,
    bargap=0.2  
)

fig.show()

# 🔍 6.- EDA (Exploratory Data Analysis) Analisis Bivariante

En esta parte intentaremos responder a algunas preguntas relacionadas con la relación entre 2 variables.

Los objetivos de este EDA bivariante van a ser los siguientes:

1) Analizar el Loan status y sus relacionados

2) Analizar los Loan Ammount y sus relacionados

3) Analizar los Purpose y sus relacionados

4) Analizar los Interest Rate y sus relacionados

Esto servirá para comprender qué variables son importantes, ver estadísticas resumidas y visualizar los datos.

## 6.1.- Loan status y relacionados

### 6.1.1.- loan_status y term

In [80]:
crosstab_loan_term = pd.crosstab(data['loan_status'], data['term'], normalize="index").mul(100)
crosstab_loan_term

term                                                36 months  60 months
loan_status                                                             
Charged Off                                         60.045015  39.954985
Current                                             62.611637  37.388363
Default                                             70.900693  29.099307
Does not meet the credit policy. Status:Charged...  85.282523  14.717477
Does not meet the credit policy. Status:Fully Paid  89.989940  10.010060
Fully Paid                                          78.489808  21.510192
In Grace Period                                     57.768249  42.231751
Issued                                              76.285160  23.714840
Late (16-30 days)                                   60.684075  39.315925
Late (31-120 days)                                  58.691346  41.308654

In [ ]:
fig = go.Figure()

for term in crosstab_loan_term.columns:
    fig.add_trace(go.Bar(
        x=crosstab_loan_term.index,
        y=crosstab_loan_term[term],
        name=term,
        text=crosstab_loan_term[term].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig.update_layout(
    title='Loan status by term',
    xaxis_title='Loan Status',
    yaxis_title='Percentage (%)',
    title_font_size=16,
    barmode='group',  
    bargap=0.2,  
    width=900,
    height=600
)

fig.show()

### 6.1.2.- loan_status y grade

In [82]:
crosstab_loan_grade = pd.crosstab(data['loan_status'], data['grade'], normalize='index').mul(100)
crosstab_loan_grade

grade                                                       A          B  \
loan_status                                                                
Charged Off                                          6.164425  20.374130   
Current                                             29.588290  29.370543   
Default                                              7.390300  27.251732   
Does not meet the credit policy. Status:Charged...   1.051248  11.169514   
Does not meet the credit policy. Status:Fully Paid   4.527163  13.531187   
Fully Paid                                          21.629836  31.562449   
In Grace Period                                     13.911049  24.660949   
Issued                                              58.147430  22.211445   
Late (16-30 days)                                   13.497609  24.310408   
Late (31-120 days)                                  11.353225  25.324997   

grade                                                       C          D  \
loan_status                                                                
Charged Off                                         32.568929  22.824840   
Current                                             25.561485  12.967791   
Default                                             36.489607  20.092379   
Does not meet the credit policy. Status:Charged...  19.448095  25.886991   
Does not meet the credit policy. Status:Fully Paid  24.195171  24.849095   
Fully Paid                                          27.375728  12.835638   
In Grace Period                                     32.548863  22.506980   
Issued                                              14.645975   4.995150   
Late (16-30 days)                                   34.056638  22.361162   
Late (31-120 days)                                  33.124923  23.727869   

grade                                                       E          F  \
loan_status                                                                
Charged Off                                         12.079504   4.542847   
Current                                              1.992889   0.403098   
Default                                              7.159353   1.616628   
Does not meet the credit policy. Status:Charged...  20.762155  12.220762   
Does not meet the credit policy. Status:Fully Paid  19.014085   7.746479   
Fully Paid                                           4.854241   1.377903   
In Grace Period                                      4.736737   1.256482   
Issued                                               0.000000   0.000000   
Late (16-30 days)                                    4.597278   0.919456   
Late (31-120 days)                                   5.088523   1.033800   

grade                                                      G  
loan_status                                                   
Charged Off                                         1.445326  
Current                                             0.115905  
Default                                             0.000000  
Does not meet the credit policy. Status:Charged...  9.461235  
Does not meet the credit policy. Status:Fully Paid  6.136821  
Fully Paid                                          0.364205  
In Grace Period                                     0.378939  
Issued                                              0.000000  
Late (16-30 days)                                   0.257448  
Late (31-120 days)                                  0.346663

In [ ]:
fig = go.Figure()

for grade in crosstab_loan_grade.columns:
    fig.add_trace(go.Bar(
        x=crosstab_loan_grade.index,
        y=crosstab_loan_grade[grade],
        name=grade,
        text=crosstab_loan_grade[grade].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig.update_layout(
    title='Loan status by grade',
    xaxis_title='Loan Status',
    yaxis_title='Percentage (%)',
    title_font_size=16,
    barmode='group',  
    bargap=0.2,  
    width=1500,
    height=600,
    legend_title_text='Grade',  
    legend=dict(
        title_font_family="Arial",
        font=dict(
            family="Arial",
            size=12,
            color="black"
        )
    )
)

fig.show()

### 6.1.3.- loan_status y home_ownership

In [84]:
crosstab_loan_home_ownership = pd.crosstab(data['loan_status'], data['home_ownership'], normalize='index').mul(100)
crosstab_loan_home_ownership

home_ownership                                           ANY   MORTGAGE  \
loan_status                                                               
Charged Off                                         0.065922  42.619184   
Current                                             0.219783  48.924847   
Default                                             0.230947  44.110855   
Does not meet the credit policy. Status:Charged...  0.000000  45.729304   
Does not meet the credit policy. Status:Fully Paid  0.000000  45.674044   
Fully Paid                                          0.062426  50.976210   
In Grace Period                                     0.219386  48.923016   
Issued                                              0.000000  48.351115   
Late (16-30 days)                                   0.220669  45.972784   
Late (31-120 days)                                  0.309521  42.144361   

home_ownership                                          NONE     OTHER  \
loan_status                                                              
Charged Off                                         0.001931  0.007447   
Current                                             0.000097  0.000000   
Default                                             0.000000  0.000000   
Does not meet the credit policy. Status:Charged...  0.131406  1.445466   
Does not meet the credit policy. Status:Fully Paid  0.201207  1.358149   
Fully Paid                                          0.002804  0.007812   
In Grace Period                                     0.000000  0.000000   
Issued                                              0.000000  0.000000   
Late (16-30 days)                                   0.000000  0.000000   
Late (31-120 days)                                  0.000000  0.000000   

home_ownership                                            OWN       RENT  
loan_status                                                               
Charged Off                                         11.401801  45.903715  
Current                                             11.786820  39.068453  
Default                                             14.087760  41.570439  
Does not meet the credit policy. Status:Charged...   6.438896  46.254928  
Does not meet the credit policy. Status:Fully Paid   6.941650  45.824950  
Fully Paid                                          10.965607  37.985142  
In Grace Period                                     12.704428  38.153171  
Issued                                              12.075655  39.573230  
Late (16-30 days)                                   12.063259  41.743288  
Late (31-120 days)                                  12.876068  44.670051

In [ ]:
fig = go.Figure()

for home_ownership in crosstab_loan_home_ownership.columns:
    fig.add_trace(go.Bar(
        x=crosstab_loan_home_ownership.index,
        y=crosstab_loan_home_ownership[home_ownership],
        name=home_ownership,
        text=crosstab_loan_home_ownership[home_ownership].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig.update_layout(
    title='Loan status by home ownership',
    xaxis_title='Loan Status',
    yaxis_title='Percentage (%)',
    title_font_size=16,
    barmode='group',  
    bargap=0.2,  
    width=1500,
    height=600,
    legend_title_text='Home Ownership',  
    legend=dict(
        title_font_family="Arial",
        font=dict(
            family="Arial",
            size=12,
            color="black"
        ),
        x=1.1,  
        y=0.5,  
        xanchor='left',
        yanchor='middle'
    )
)

fig.show()

### 6.1.4.- loan_status y verification_status

In [86]:
crosstab_loan_verification_status = pd.crosstab(data['loan_status'], data['verification_status'], normalize='index').mul(100)
crosstab_loan_verification_status

verification_status                                 Not Verified  \
loan_status                                                        
Charged Off                                            24.550404   
Current                                                44.942077   
Default                                                30.023095   
Does not meet the credit policy. Status:Charged...     67.148489   
Does not meet the credit policy. Status:Fully Paid     66.448692   
Fully Paid                                             33.339543   
In Grace Period                                        36.507778   
Issued                                                 60.814743   
Late (16-30 days)                                      38.727473   
Late (31-120 days)                                     37.365358   

verification_status                                 Source Verified   Verified  
loan_status                                                                     
Charged Off                                               40.792943  34.656652  
Current                                                   38.455465  16.602458  
Default                                                   41.801386  28.175520  
Does not meet the credit policy. Status:Charged...        10.775296  22.076216  
Does not meet the credit policy. Status:Fully Paid        10.462777  23.088531  
Fully Paid                                                39.136110  27.524348  
In Grace Period                                           40.147587  23.344635  
Issued                                                    32.977692   6.207565  
Late (16-30 days)                                         38.249356  23.023170  
Late (31-120 days)                                        40.113904  22.520738

In [ ]:
fig = go.Figure()

for verification_status in crosstab_loan_verification_status.columns:
    fig.add_trace(go.Bar(
        x=crosstab_loan_verification_status.index,
        y=crosstab_loan_verification_status[verification_status],
        name=verification_status,
        text=crosstab_loan_verification_status[verification_status].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig.update_layout(
    title='Loan status by verification status',
    xaxis_title='Loan Status',
    yaxis_title='Percentage (%)',
    title_font_size=16,
    barmode='group',  
    bargap=0.2,  
    width=1200,
    height=500,
    legend_title_text='Verification Status',  
    legend=dict(
        title_font_family="Arial",
        font=dict(
            family="Arial",
            size=16,  
            color="black"
        ),
        x=1.05,  
        y=0.5,  
        xanchor='left',
        yanchor='middle'
    )
)

fig.show()

### 6.1.5.- loan_status con annual_inc_range, emp_length, y purpose

Loan status and annual income range

In [88]:
crosstab_annual_income = pd.crosstab(data['loan_status'], data['annual_inc_range'], normalize='index').mul(100)
crosstab_annual_income

annual_inc_range                                           0  0.1-20000  \
loan_status                                                               
Charged Off                                         0.059578   1.771627   
Current                                             0.185642   1.648859   
Default                                             0.230947   1.616628   
Does not meet the credit policy. Status:Charged...  0.000000   5.781866   
Does not meet the credit policy. Status:Fully Paid  0.000000   4.889113   
Fully Paid                                          0.058019   1.286101   
In Grace Period                                     0.189469   1.785002   
Issued                                              0.096993   0.921435   
Late (16-30 days)                                   0.183891   2.721589   
Late (31-120 days)                                  0.179522   2.296645   

annual_inc_range                                    20000-40000  40000-60000  \
loan_status                                                                    
Charged Off                                           16.539603    27.883480   
Current                                               12.342291    22.812740   
Default                                               15.242494    21.709007   
Does not meet the credit policy. Status:Charged...    22.339028    27.332457   
Does not meet the credit policy. Status:Fully Paid    21.925403    25.604839   
Fully Paid                                            13.542482    25.454689   
In Grace Period                                       12.634623    22.776227   
Issued                                                 9.505335    20.417071   
Late (16-30 days)                                     14.159618    21.772711   
Late (31-120 days)                                    14.727003    24.235483   

annual_inc_range                                    60000-80000  80000-100000  \
loan_status                                                                     
Charged Off                                           22.934894     13.398226   
Current                                               21.907032     14.775522   
Default                                               25.866051     12.702079   
Does not meet the credit policy. Status:Charged...    19.842313      8.409987   
Does not meet the credit policy. Status:Fully Paid    17.792339     12.096774   
Fully Paid                                            22.671709     14.696254   
In Grace Period                                       21.569605     15.027922   
Issued                                                22.502425     15.809893   
Late (16-30 days)                                     19.970577     14.674513   
Late (31-120 days)                                    21.740745     14.114151   

annual_inc_range                                    100000-120000  \
loan_status                                                         
Charged Off                                              6.760208   
Current                                                  8.804034   
Default                                                  8.775982   
Does not meet the credit policy. Status:Charged...       4.599212   
Does not meet the credit policy. Status:Fully Paid       6.451613   
Fully Paid                                               8.373910   
In Grace Period                                          8.436378   
Issued                                                   9.359845   
Late (16-30 days)                                        8.753218   
Late (31-120 days)                                       7.738022   

annual_inc_range                                    120000-140000  \
loan_status                                                         
Charged Off                                              4.354734   
Current                                                  6.240543   
Default                                                  6.6

In [ ]:
fig = go.Figure()

for annual_inc_range in crosstab_annual_income.columns:
    fig.add_trace(go.Bar(
        x=crosstab_annual_income.index,
        y=crosstab_annual_income[annual_inc_range],
        name=annual_inc_range,
        text=crosstab_annual_income[annual_inc_range].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig.update_layout(
    title='Annual income range by Loan status',
    xaxis_title='Loan Status',
    yaxis_title='Percentage (%)',
    title_font_size=16,
    barmode='group',  
    bargap=0.2,  
    width=1500,
    height=600,
    legend_title_text='Annual Income Range',  
    legend=dict(
        title_font_family="Arial",
        font=dict(
            family="Arial",
            size=12,  
            color="black"
        ),
        x=1.2,  
        y=0.5,  
        xanchor='left',
        yanchor='middle'
    )
)

fig.show()

loan status and employment length

In [90]:
crosstab_emp_length = pd.crosstab(data['loan_status'], data['emp_length'], normalize='index').mul(100)
crosstab_emp_length

emp_length                                             1 year  10+ years  \
loan_status                                                                
Charged Off                                          7.438258  33.385972   
Current                                              7.275244  34.352349   
Default                                             10.840108  35.772358   
Does not meet the credit policy. Status:Charged...  12.198391  20.911528   
Does not meet the credit policy. Status:Fully Paid  13.052311  15.947181   
Fully Paid                                           6.961050  35.466836   
In Grace Period                                      7.010108  33.855016   
Issued                                               8.088235  32.825630   
Late (16-30 days)                                    7.616602  34.403081   
Late (31-120 days)                                   7.381518  32.951209   

emp_length                                            2 years    3 years  \
loan_status                                                                
Charged Off                                          9.772018   8.791381   
Current                                              9.550515   8.491871   
Default                                              7.859079  10.840108   
Does not meet the credit policy. Status:Charged...  11.126005   9.651475   
Does not meet the credit policy. Status:Fully Paid  13.509396   9.852717   
Fully Paid                                           9.642814   8.509753   
In Grace Period                                      9.172916   8.944680   
Issued                                               9.558824   8.298319   
Late (16-30 days)                                    8.515190   8.001712   
Late (31-120 days)                                   9.646302   8.388089   

emp_length                                           4 years   5 years  \
loan_status                                                              
Charged Off                                         6.601070  6.750546   
Current                                             6.531932  6.809538   
Default                                             6.775068  5.962060   
Does not meet the credit policy. Status:Charged...  7.506702  6.702413   
Does not meet the credit policy. Status:Fully Paid  7.567293  6.196039   
Fully Paid                                          6.398807  6.623067   
In Grace Period                                     6.521030  6.597109   
Issued                                              7.300420  6.827731   
Late (16-30 days)                                   7.787762  7.017544   
Late (31-120 days)                                  6.815322  6.731441   

emp_length                                           6 years   7 years  \
loan_status                                                              
Charged Off                                         4.849243  4.475552   
Current                                             4.642795  3.887119   
Default                                             4.065041  3.252033   
Does not meet the credit policy. Status:Charged...  5.764075  4.289544   
Does not meet the credit policy. Status:Fully Paid  5.129507  3.453530   
Fully Paid                                          4.907710  4.484486   
In Grace Period                                     4.619063  3.673514   
Issued                                              4.831933  4.306723   
Late (16-30 days)                                   4.749679  3.851091   
Late (31-120 days)                                  4.690340  3.662799   

emp_length                                           8 years   9 years  \
loan_status                                                              
Charged Off                                         4.565057  3.825510   
Current                                             3.483972  2.887230   
Default                                             3.523035  2.439024   
Does not meet the credit p

In [ ]:
fig = go.Figure()

for emp_length in crosstab_emp_length.columns:
    fig.add_trace(go.Bar(
        x=crosstab_emp_length.index,
        y=crosstab_emp_length[emp_length],
        name=emp_length,
        text=crosstab_emp_length[emp_length].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig.update_layout(
    title='Employment length by Loan status',
    xaxis_title='Loan Status',
    yaxis_title='Percentage (%)',
    title_font_size=16,
    barmode='group',  
    bargap=0.2,  
    width=1500,
    height=600,
    legend_title_text='Employment Length',  
    legend=dict(
        title_font_family="Arial",
        font=dict(
            family="Arial",
            size=12,  
            color="black"
        ),
        x=1.1,  
        y=0.5,  
        xanchor='left',
        yanchor='middle'
    )
)

fig.show()

loan status and purposes

In [92]:
crosstab_purpose = pd.crosstab(data['loan_status'], data['purpose'], normalize='index').mul(100)
crosstab_purpose

purpose                                                  car  credit_card  \
loan_status                                                                 
Charged Off                                         0.787206    19.190011   
Current                                             0.902023    26.710255   
Default                                             0.923788    19.168591   
Does not meet the credit policy. Status:Charged...  1.708279     9.067017   
Does not meet the credit policy. Status:Fully Paid  2.565392    13.631791   
Fully Paid                                          1.159113    22.918607   
In Grace Period                                     0.588353    21.559633   
Issued                                              0.678952    28.855480   
Late (16-30 days)                                   1.029790    22.066936   
Late (31-120 days)                                  0.736660    22.619785   

purpose                                             debt_consolidation  \
loan_status                                                              
Charged Off                                                  60.545362   
Current                                                      53.876080   
Default                                                      58.891455   
Does not meet the credit policy. Status:Charged...           38.370565   
Does not meet the credit policy. Status:Fully Paid           40.643863   
Fully Paid                                                   56.346814   
In Grace Period                                              58.526127   
Issued                                                       52.473327   
Late (16-30 days)                                            56.123575   
Late (31-120 days)                                           57.546119   

purpose                                             educational  \
loan_status                                                       
Charged Off                                            0.015446   
Current                                                0.000000   
Default                                                0.000000   
Does not meet the credit policy. Status:Charged...     4.204993   
Does not meet the credit policy. Status:Fully Paid     3.269618   
Fully Paid                                             0.018093   
In Grace Period                                        0.000000   
Issued                                                 0.000000   
Late (16-30 days)                                      0.000000   
Late (31-120 days)                                     0.000000   

purpose                                             home_improvement  \
loan_status                                                            
Charged Off                                                 5.883359   
Current                                                     6.493207   
Default                                                     6.235566   
Does not meet the credit policy. Status:Charged...          9.329829   
Does not meet the credit policy. Status:Fully Paid          7.193159   
Fully Paid                                                  6.845518   
In Grace Period                                             6.631432   
Issued                                                      7.225994   
Late (16-30 days)                                           5.847738   
Late (31-120 days)                                          6.060418   

purpose                                                house  major_purchase  \
loan_status                                                                    
Charged Off                                         0.678807        2.126615   
Current                                             0.790967        1.953413   
Default                                             2.078522        2.309469   
Does not meet the credit policy. Status:Charged...  1.445466        3.022339   
Does not meet the credit policy. Sta

In [ ]:
fig = go.Figure()

for purpose in crosstab_purpose.columns:
    fig.add_trace(go.Bar(
        x=crosstab_purpose.index,
        y=crosstab_purpose[purpose],
        name=purpose,
        text=crosstab_purpose[purpose].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig.update_layout(
    title='Purpose of loan in each loan status',
    xaxis_title='Loan Status',
    yaxis_title='Percentage (%)',
    title_font_size=16,
    barmode='group',  
    bargap=0.2,  
    width=1500,
    height=600,
    legend_title_text='Loan Purpose',  
    legend=dict(
        title_font_family="Arial",
        font=dict(
            family="Arial",
            size=12,  
            color="black"
        ),
        x=1.2,  
        y=0.5,  
        xanchor='left',
        yanchor='middle'
    )
)

fig.show()

un vistazo más de cerca sólo al estado Charged Off

In [ ]:
fig, axs = plt.subplots(1, 3, figsize=(20, 7), sharex=True)
pd.crosstab(data['loan_status'], data['annual_inc_range'], normalize='index').filter(items=['Charged Off'], axis=0).plot(kind='bar',ax=axs[0])
pd.crosstab(data['loan_status'], data['emp_length'], normalize='index').filter(items=['Charged Off'], axis=0).plot(kind='bar',ax=axs[1])
pd.crosstab(data['loan_status'], data['purpose'], normalize='index').filter(items=['Charged Off'], axis=0).plot(kind='bar',ax=axs[2])
fig.suptitle('Distribution of Annual income range, Employment length and Purpose in Charged Off status (respectively)', fontsize=16)

### 6.1.6.- loan_status y open_acc/ total_acc

In [116]:
df_loan_open_1 = pd.crosstab(data['loan_status'], data['open_acc'], normalize='index').mul(100)

df_loan_open_1 = df_loan_open_1.loc[:, (df_loan_open_1 != 0).any(axis=0)]

Gráfico 1: Percentage of Charged Off status by the number of open credit lines in the borrower's credit file

In [117]:
df_loan_open_1_sorted_charged_off = df_loan_open_1.sort_values(axis=1, by='Charged Off', ascending=False).iloc[0, :30]

In [ ]:
fig1 = go.Figure()

fig1.add_trace(go.Bar(
    x=df_loan_open_1_sorted_charged_off.index,
    y=df_loan_open_1_sorted_charged_off.values,
    text=df_loan_open_1_sorted_charged_off.round(2).astype(str) + '%',
    textposition='outside',
    marker_color='blue'
))

fig1.update_layout(
    title="Top 30: Percentage of Charged Off status by the number of open credit lines in the borrower's credit file",
    xaxis_title="Number of Open Credit Lines",
    yaxis_title="Percentage (%)",
    title_font_size=16,
    width=1200,
    height=500
)

fig1.show()

Gráfico 2: Percentage of Fully Paid loan status by the number of open credit lines

In [119]:
df_loan_open_1_sorted_fully_paid = df_loan_open_1.sort_values(axis=1, by='Fully Paid', ascending=False).iloc[5, :30]

In [ ]:
fig2 = go.Figure()

fig2.add_trace(go.Bar(
    x=df_loan_open_1_sorted_fully_paid.index,
    y=df_loan_open_1_sorted_fully_paid.values,
    text=df_loan_open_1_sorted_fully_paid.round(2).astype(str) + '%',
    textposition='outside',
    marker_color='yellow'
))

fig2.update_layout(
    title="Top 30: Percentage of Fully Paid loan status by the number of open credit lines",
    xaxis_title="Number of Open Credit Lines",
    yaxis_title="Percentage (%)",
    title_font_size=16,
    width=1200,
    height=500
)

fig2.show()

In [121]:
df_loan_open_2 = pd.crosstab(data['loan_status'], data['total_acc'], normalize='index').mul(100)

df_loan_open_2 = df_loan_open_2.loc[:, (df_loan_open_2 != 0).any(axis=0)]

Gráfico 3: Percentage of Charged Off loan status by the total number of credit lines currently in the borrower's credit file

In [123]:
df_loan_open_2_sorted_charged_off = df_loan_open_2.sort_values(axis=1, by='Charged Off', ascending=False).iloc[0, :30]

In [ ]:
fig3 = go.Figure()

fig3.add_trace(go.Bar(
    x=df_loan_open_2_sorted_charged_off.index,
    y=df_loan_open_2_sorted_charged_off.values,
    text=df_loan_open_2_sorted_charged_off.round(2).astype(str) + '%',
    textposition='outside',
    marker_color='green'
))

fig3.update_layout(
    title="Top 30: Percentage of Charged Off loan status by the total number of credit lines currently in the borrower's credit file",
    xaxis_title="Total Number of Credit Lines",
    yaxis_title="Percentage (%)",
    title_font_size=16,
    width=1200,
    height=500
)

fig3.show()

Gráfico 4: Percentage of Fully Paid loan status by the total number of credit lines

In [125]:
df_loan_open_2_sorted_fully_paid = df_loan_open_2.sort_values(axis=1, by='Fully Paid', ascending=False).iloc[5, :30]

In [ ]:
fig4 = go.Figure()

fig4.add_trace(go.Bar(
    x=df_loan_open_2_sorted_fully_paid.index,
    y=df_loan_open_2_sorted_fully_paid.values,
    text=df_loan_open_2_sorted_fully_paid.round(2).astype(str) + '%',
    textposition='outside',
    marker_color='red'
))

fig4.update_layout(
    title="Top 30: Percentage of Fully Paid loan status by the total number of credit lines",
    xaxis_title="Total Number of Credit Lines",
    yaxis_title="Percentage (%)",
    title_font_size=16,
    width=1200,
    height=500
)

fig4.show()

### 6.1.7.- loan_status y pub_rec/ pub_rec_bankruptcies

loan status and number of derogatory public records

Gráfico 1: Number of derogatory public records by loan status

In [133]:
loan_status_pub_rec = pd.crosstab(data['loan_status'], data['pub_rec'])

In [ ]:
fig1 = go.Figure()

for col in loan_status_pub_rec.columns:
    fig1.add_trace(go.Bar(
        y=loan_status_pub_rec.index,
        x=loan_status_pub_rec[col],
        name=str(col),
        orientation='h'  
    ))

fig1.update_layout(
    title='Number of derogatory public records by loan status',
    yaxis_title='Loan Status',
    xaxis_title='Frequency',
    barmode='stack',
    title_font_size=16,
    width=1200,
    height=600
)

fig1.show()

Gráfico 2: Number of public record bankruptcies by loan status

In [135]:
loan_status_pub_rec_bankruptcies = pd.crosstab(data['loan_status'], data['pub_rec_bankruptcies'])

In [ ]:
fig2 = go.Figure()

for col in loan_status_pub_rec_bankruptcies.columns:
    fig2.add_trace(go.Bar(
        y=loan_status_pub_rec_bankruptcies.index,
        x=loan_status_pub_rec_bankruptcies[col],
        name=str(col),
        orientation='h'  
    ))

fig2.update_layout(
    title='Number of public record bankruptcies by loan status',
    yaxis_title='Loan Status',
    xaxis_title='Frequency',
    barmode='stack',
    title_font_size=16,
    width=1200,
    height=600
)

fig2.show()

### 6.1.8.- Loan status y interest rate / loan amount

Gráfico 1: Loan Status by Interest Rate

In [ ]:
fig1 = go.Figure()

colors = px.colors.sequential.Viridis

for i, status in enumerate(data['loan_status'].unique()):
    subset = data[data['loan_status'] == status]
    fig1.add_trace(go.Box(
        y=subset['int_rate'],
        name=status,
        marker_color=colors[i % len(colors)]
    ))

fig1.update_layout(
    title='Loan Status by Interest Rate',
    xaxis_title="Loan Status",
    yaxis_title="Interest Rate",
    title_font_size=16,
    width=1200,
    height=600
)

fig1.show()

Gráfico 2: Loan Status by Loan Amount

In [ ]:
fig2 = go.Figure()

for i, status in enumerate(data['loan_status'].unique()):
    subset = data[data['loan_status'] == status]
    fig2.add_trace(go.Box(
        y=subset['loan_amnt'],
        name=status,
        marker_color=colors[i % len(colors)]
    ))

fig2.update_layout(
    title='Loan Status by Loan Amount',
    xaxis_title="Loan Status",
    yaxis_title="Loan Amount",
    title_font_size=16,
    width=1200,
    height=600
)

fig2.show()

### 6.1.9.- Loan status per state

In [147]:
df_loan_status_state = pd.crosstab(data['loan_status'], data['addr_state'], normalize='index').mul(100)

Gráfico 1: Percentage of Charged Off status per state

In [152]:
df_charged_off = df_loan_status_state.sort_values(axis=1, by='Charged Off', ascending=False).iloc[0, :]

In [ ]:
fig1 = go.Figure()

fig1.add_trace(go.Bar(
    x=df_charged_off.index,
    y=df_charged_off.values,
    text=df_charged_off.round(2).astype(str) + '%',
    textposition='outside',
    marker=dict(color=df_charged_off.values, colorscale='Viridis')
))

fig1.update_layout(
    title='Percentage of Charged Off status per state',
    xaxis_title='State',
    yaxis_title='Percentage (%)',
    title_font_size=16,
    width=1200,
    height=600
)

fig1.show()

Gráfico 2: Percentage of Fully Paid status per state

In [155]:
df_fully_paid = df_loan_status_state.sort_values(axis=1, by='Fully Paid', ascending=False).iloc[5, :]

In [ ]:
fig2 = go.Figure()

fig2.add_trace(go.Bar(
    x=df_fully_paid.index,
    y=df_fully_paid.values,
    text=df_fully_paid.round(2).astype(str) + '%',
    textposition='outside',
    marker=dict(color=df_fully_paid.values, colorscale='Viridis')
))

fig2.update_layout(
    title='Percentage of Fully Paid status per state',
    xaxis_title='State',
    yaxis_title='Percentage (%)',
    title_font_size=16,
    width=1200,
    height=600
)

fig2.show()

## 6.2.- Loan ammount

### 6.2.1.- Loan amount vs year experience

In [158]:
df_loan_ammount_range_emp = pd.crosstab(data['loan_amnt_range_5000'], data['emp_length'], normalize='index').mul(100)
df_loan_ammount_range_emp

emp_length              1 year  10+ years    2 years   3 years   4 years  \
loan_amnt_range_5000                                                       
0-5000                8.518763  30.416259  10.858916  9.264026  7.017594   
5000-10000            8.206148  30.645514  10.726578  9.369164  6.959260   
10000-15000           7.418744  33.001907  10.000048  8.915728  6.738499   
15000-20000           6.714565  35.760072   9.276963  8.372202  6.371710   
20000-25000           6.345544  38.558770   8.617654  7.801212  6.016872   
25000-30000           5.941474  40.332779   8.405620  7.453662  5.810356   
30000-35000           5.888303  39.940633   8.522427  7.545441  5.787892   
35000-40000           5.735733  41.345218   8.102597  7.362744  5.672484   

emp_length             5 years   6 years   7 years   8 years   9 years  \
loan_amnt_range_5000                                                     
0-5000                6.899318  4.834188  4.066006  3.871607  3.245438   
5000-10000            7.012924  4.952724  4.257225  4.069401  3.400555   
10000-15000           6.953263  4.922868  4.368618  4.204602  3.573355   
15000-20000           6.708012  4.843492  4.476767  4.323477  3.710317   
20000-25000           6.440664  4.799853  4.303590  4.271883  3.579153   
25000-30000           6.234592  4.579909  4.143753  4.145378  3.603028   
30000-35000           6.226180  4.488420  4.067722  4.030343  3.466725   
35000-40000           5.917992  4.345076  4.065447  4.108723  3.434616   

emp_length             < 1 year  
loan_amnt_range_5000             
0-5000                11.007886  
5000-10000            10.400507  
10000-15000            9.902370  
15000-20000            9.442423  
20000-25000            9.264806  
25000-30000            9.349450  
30000-35000           10.035913  
35000-40000            9.909370

In [ ]:
fig = go.Figure()

for col in df_loan_ammount_range_emp.columns:
    fig.add_trace(go.Bar(
        x=df_loan_ammount_range_emp.index,
        y=df_loan_ammount_range_emp[col],
        name=col,
        text=df_loan_ammount_range_emp[col].round(2).astype(str) + '%',
        textposition='outside'
    ))

fig.update_layout(
    title='Loan Amount Range by Employment Length',
    xaxis_title='Loan Amount Range',
    yaxis_title='Percentage (%)',
    barmode='stack',
    title_font_size=16,
    width=1200,
    height=600
)

fig.show()

### 6.2.2.- Loan amount y term

In [ ]:
fig = px.box(data, x='term', y='loan_amnt', title='Loan amount distribution by term', color_discrete_sequence=['#1f77b4'])

fig.update_layout(
    xaxis_title="Term",
    yaxis_title="Loan Amount",
    title_font_size=16,
    width=1200,
    height=600
)

fig.show()

### 6.2.3.- Loan amount y purpose

In [165]:
df_purpose_loan_amnt_range = pd.crosstab(data['purpose'], data['loan_amnt_range_5000'], normalize='index').mul(100)

In [ ]:
fig = go.Figure()

for col in df_purpose_loan_amnt_range.columns:
    fig.add_trace(go.Bar(
        x=df_purpose_loan_amnt_range.index,
        y=df_purpose_loan_amnt_range[col],
        name=col,
        text=df_purpose_loan_amnt_range[col].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig.update_layout(
    title='Loan amount by Purpose',
    xaxis_title='Purpose',
    yaxis_title='Percentage (%)',
    barmode='stack',
    title_font_size=16,
    width=1200,
    height=600
)

fig.show()

In [ ]:
sns.boxplot(x='purpose', y='loan_amnt', data=data, palette="Accent")
plt.title("Loan amount distribution by purpose", fontsize=16)
plt.ylabel('Loan amount in USD')
plt.tick_params(axis='x', rotation=90)

## 6.3.- Purpose

### 6.3.1.- Purpose y term

In [169]:
df_purpose_term = pd.crosstab(data['purpose'], data['term'], normalize='index').mul(100)

In [ ]:
fig = go.Figure()

for col in df_purpose_term.columns:
    fig.add_trace(go.Bar(
        x=df_purpose_term.index,
        y=df_purpose_term[col],
        name=col,
        text=df_purpose_term[col].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig.update_layout(
    title='Loan term by Purpose',
    xaxis_title='Purpose',
    yaxis_title='Percentage (%)',
    barmode='stack',
    title_font_size=16,
    width=1200,
    height=600
)

fig.show()

### 6.3.2.- Purpose y grade

In [172]:
df_purpose_grade = pd.crosstab(data['purpose'], data['grade'], normalize='index').mul(100)

In [ ]:
fig = go.Figure()

for col in df_purpose_grade.columns:
    fig.add_trace(go.Bar(
        x=df_purpose_grade.index,
        y=df_purpose_grade[col],
        name=col,
        text=df_purpose_grade[col].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig.update_layout(
    title='Grade by Purpose',
    xaxis_title='Purpose',
    yaxis_title='Percentage (%)',
    barmode='stack',
    title_font_size=16,
    width=1200,
    height=600,
    legend=dict(
        title='Grade',
        x=1.1,
        y=0.5,
        traceorder='normal'
    )
)

fig.show()

## 6.4.- Interest Rate

### 6.4.1.- Interest rate y annual income

Gráfico 1: Annual income range by Interest rate range

In [176]:
df_interest_annual_inc = pd.crosstab(data['interest_range'], data['annual_inc_range'], normalize='index').mul(100)

In [ ]:
fig1 = go.Figure()

for col in df_interest_annual_inc.columns:
    fig1.add_trace(go.Bar(
        x=df_interest_annual_inc.index,
        y=df_interest_annual_inc[col],
        name=col,
        text=df_interest_annual_inc[col].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig1.update_layout(
    title='Annual income range by Interest rate range',
    xaxis_title='Interest Rate Range',
    yaxis_title='Percentage (%)',
    barmode='stack',
    title_font_size=16,
    width=1200,
    height=600,
    legend=dict(
        title='Annual Income Range',
        x=1.2,
        y=0.5,
        traceorder='normal'
    )
)

fig1.show()

Gráfico 2: Loan amount range by Interest rate range

In [178]:
df_interest_loan_amnt = pd.crosstab(data['interest_range'], data['loan_amnt_range_5000'], normalize='index').mul(100)

In [ ]:
fig2 = go.Figure()

for col in df_interest_loan_amnt.columns:
    fig2.add_trace(go.Bar(
        x=df_interest_loan_amnt.index,
        y=df_interest_loan_amnt[col],
        name=col,
        text=df_interest_loan_amnt[col].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig2.update_layout(
    title='Loan amount range by Interest rate range',
    xaxis_title='Interest Rate Range',
    yaxis_title='Percentage (%)',
    barmode='stack',
    title_font_size=16,
    width=1200,
    height=600,
    legend=dict(
        title='Loan Amount Range',
        x=1.2,
        y=0.5,
        traceorder='normal'
    )
)

fig2.show()

### 6.4.2.- Interest rate y term

Gráfico 1: Loan term by Interest rate range

In [181]:
df_interest_term = pd.crosstab(data['interest_range'], data['term'], normalize='index').mul(100)

In [ ]:
fig1 = go.Figure()

for col in df_interest_term.columns:
    fig1.add_trace(go.Bar(
        x=df_interest_term.index,
        y=df_interest_term[col],
        name=col,
        text=df_interest_term[col].round(2).astype(str) + '%',
        textposition='outside'
    ))

fig1.update_layout(
    title='Loan term by Interest rate range',
    xaxis_title='Interest Rate Range',
    yaxis_title='Percentage (%)',
    barmode='group',
    title_font_size=16,
    width=1200,
    height=600,
    legend=dict(
        title='Loan Term',
        x=1.2,
        y=0.5,
        traceorder='normal'
    )
)

fig1.show()

Gráfico 2: Loan amount range by Interest rate range

In [183]:
df_interest_loan_amnt = pd.crosstab(data['interest_range'], data['loan_amnt_range_5000'], normalize='index').mul(100)

In [ ]:
fig2 = go.Figure()

for col in df_interest_loan_amnt.columns:
    fig2.add_trace(go.Bar(
        x=df_interest_loan_amnt.index,
        y=df_interest_loan_amnt[col],
        name=col,
        text=df_interest_loan_amnt[col].round(2).astype(str) + '%',
        textposition='inside'
    ))

fig2.update_layout(
    title='Loan amount range by Interest rate range',
    xaxis_title='Interest Rate Range',
    yaxis_title='Percentage (%)',
    barmode='stack',
    title_font_size=16,
    width=1200,
    height=600,
    legend=dict(
        title='Loan Amount Range',
        x=1.2,
        y=0.5,
        traceorder='normal'
    )
)

fig2.show()

### 6.4.3.- Interest vs sub-grade

In [ ]:
fig = px.box(data, x='grade', y='int_rate', title='Interest rate distribution by Grade', color='grade', color_discrete_sequence=px.colors.qualitative.Pastel)

fig.update_layout(
    xaxis_title="Grade",
    yaxis_title="Interest Rate",
    title_font_size=16,
    width=1200,
    height=600
)

fig.show()